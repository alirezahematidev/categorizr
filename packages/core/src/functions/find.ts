import { assertion, findNode, nonUniqueTreeWarning } from "../helpers";
import { MaybeNode, TreeNode } from "$core/index";

export function find<T extends TreeNode>(tree: readonly T[], id: string): MaybeNode<T> {
  assertion(tree);

  nonUniqueTreeWarning(tree, "find");

  return findNode(tree, id);
}
