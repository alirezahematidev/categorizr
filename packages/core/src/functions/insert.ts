import { Callback, TreeNode } from "$core/types";
import { clone, exception } from "../helpers";
import { findNode } from "../helpers/internal";

function insert<T extends TreeNode>(tree: readonly T[], destination: string | null, node: T): T[];
function insert<T extends TreeNode>(tree: readonly T[], destination: string | null, node: T, callback: Callback<T>): void;
function insert<T extends TreeNode>(tree: readonly T[], destination: string | null, node: T, callback?: Callback<T>) {
  const cloneTree = clone(tree);

  if (destination === null) {
    cloneTree.push(node);

    return cloneTree;
  }

  const destNode = findNode(cloneTree, destination);

  if (!destNode) throw exception("insert", "Cannot find the destination node with the given id.");

  if (!destNode.children) destNode.children = [];

  destNode.children.push(node);

  if (callback) return void callback(cloneTree);

  return cloneTree;
}

export { insert };
