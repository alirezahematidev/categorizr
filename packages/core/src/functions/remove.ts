import { Callback, TreeNode } from "$core/index";
import { assertion, clone, exception, findNode, findParent, modifyWith, nonUniqueTreeWarning } from "../helpers";

function remove<T extends TreeNode>(tree: readonly T[], id: string): T[];
function remove<T extends TreeNode>(tree: readonly T[], id: string, callback: Callback<T>): void;
function remove<T extends TreeNode>(tree: readonly T[], id: string, callback?: Callback<T>) {
  assertion(tree);

  nonUniqueTreeWarning(tree, "remove");

  const cloneTree = clone(tree);

  const node = findNode(cloneTree, id);

  if (!node) throw exception("remove", "Cannot found the node with the given id");

  const parentNode = findParent(cloneTree, node);

  if (!parentNode) {
    modifyWith(cloneTree, node.id);

    if (callback) return void callback(cloneTree);

    return cloneTree;
  }

  if (!parentNode.children) parentNode.children = [];

  modifyWith(parentNode.children, node.id);

  if (callback) return void callback(cloneTree);

  return cloneTree;
}

export { remove };
