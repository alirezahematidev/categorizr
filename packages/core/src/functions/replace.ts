import { Callback, Replacer, TreeNode } from "$core/types";
import { clone, exception, modifyWith } from "../helpers";
import { findNode, findParent } from "../helpers/internal";

function replace<T extends TreeNode>(tree: readonly T[], target: string, replacer: Replacer<T>): T[];
function replace<T extends TreeNode>(tree: readonly T[], target: string, replacer: Replacer<T>, callback: Callback<T>): void;
function replace<T extends TreeNode>(tree: readonly T[], target: string, replacer: Replacer<T>, callback?: Callback<T>) {
  const cloneTree = clone(tree);

  const targetNode = findNode(cloneTree, target);

  if (!targetNode) throw exception("replace", "Cannot found the target node with the given id");

  if (!replacer.id) replacer.id = target;

  const parentNode = findParent(cloneTree, targetNode);

  if (!parentNode) {
    modifyWith(cloneTree, targetNode.id, replacer);

    if (callback) return void callback(cloneTree);

    return cloneTree;
  }

  if (!parentNode.children) parentNode.children = [];

  modifyWith(parentNode.children, targetNode.id, replacer);

  if (callback) return void callback(cloneTree);

  return cloneTree;
}

export { replace };
