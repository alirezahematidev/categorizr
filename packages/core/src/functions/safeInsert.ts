import { CallbackWithError, TreeNode } from "$core/types";
import { clone, exception, warning } from "../helpers";
import { findNode } from "../helpers/internal";

function safeInsert<T extends TreeNode>(tree: readonly T[], destination: string | null, node: T): T[];
function safeInsert<T extends TreeNode>(tree: readonly T[], destination: string | null, node: T, callback: CallbackWithError<T>): void;
function safeInsert<T extends TreeNode>(tree: readonly T[], destination: string | null, node: T, callback?: CallbackWithError<T>) {
  const cloneTree = clone(tree);

  if (destination === null) {
    cloneTree.push(node);

    return cloneTree;
  }

  const destNode = findNode(cloneTree, destination);

  if (!destNode) {
    warning("safeInsert", "Cannot found the destination node with the given id.");

    if (callback) return void callback(tree, exception("safeInsert", "Cannot found the destination node with the given id."));

    return tree;
  }

  if (!destNode.children) destNode.children = [];

  destNode.children.push(node);

  if (callback) return void callback(cloneTree, undefined);

  return cloneTree;
}

export { safeInsert };
