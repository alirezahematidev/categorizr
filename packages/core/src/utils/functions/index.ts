export * from "./findNode";
export * from "./findParent";
export * from "./insertNode";
export * from "./safeInsertNode";
