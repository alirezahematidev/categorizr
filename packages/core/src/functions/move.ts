import { insert } from "./insert";
import { remove } from "./remove";
import { Callback, TreeNode } from "$core/index";
import { assertion, containsNode, exception, findNode, findParent, nonUniqueTreeWarning } from "../helpers";

function move<T extends TreeNode>(tree: readonly T[], from: string, to: string | null): T[];
function move<T extends TreeNode>(tree: readonly T[], from: string, to: string | null, callback: Callback<T>): void;
function move<T extends TreeNode>(tree: readonly T[], from: string, to: string | null, callback?: Callback<T>) {
  assertion(tree);

  nonUniqueTreeWarning(tree, "move");

  const node = findNode(tree, from);

  if (containsNode(tree, from, to)) throw exception("move", "Cannot move the node into its own descendants.");

  if (!node) throw exception("move", "Cannot found the source node with the given id");

  const parentNode = findParent(tree, node);

  if (parentNode && parentNode.id === to) {
    if (callback) return void callback(tree as T[]);

    return tree as T[];
  }

  const result = insert(remove(tree, from), to, node);

  if (callback) return void callback(result);

  return result;
}

export { move };
