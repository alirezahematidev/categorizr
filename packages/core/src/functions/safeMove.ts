import { safeInsert } from "./safeInsert";
import { safeRemove } from "./safeRemove";
import { CallbackWithError, TreeNode } from "$core/index";
import { containsNode, exception, findNode, findParent, error, nonUniqueTreeWarning, assertion } from "../helpers";

function safeMove<T extends TreeNode>(tree: readonly T[], from: string, to: string | null): T[];
function safeMove<T extends TreeNode>(tree: readonly T[], from: string, to: string | null, callback: CallbackWithError<T>): void;
function safeMove<T extends TreeNode>(tree: readonly T[], from: string, to: string | null, callback?: CallbackWithError<T>) {
  assertion(tree);

  nonUniqueTreeWarning(tree, "safeMove");

  const fromNode = findNode(tree, from);

  if (containsNode(tree, from, to)) {
    error("safeMove", "Cannot move the node into its own descendants.");

    if (callback) return void callback(tree, exception("safeMove", "Cannot move the node into its own descendants."));

    return [...tree];
  }

  if (!fromNode) {
    error("safeMove", "Cannot found the source node with the given id.");

    if (callback) return void callback(tree, exception("safeMove", "Cannot found the source node with the given id."));

    return [...tree];
  }

  const parentNode = findParent(tree, fromNode);

  if (parentNode && parentNode.id === to) {
    if (callback) return void callback(tree as T[], undefined);

    return [...tree] as T[];
  }

  const result = safeInsert(safeRemove(tree, from), to, fromNode);

  if (callback) return void callback(result, undefined);

  return result;
}

export { safeMove };
