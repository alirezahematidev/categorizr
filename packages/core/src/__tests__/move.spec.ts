import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";
import { move } from "../functions";
import { TreeNode } from "$core/index";

describe("move", async () => {
  beforeAll(() => {
    vi.fn().mockImplementation(move);
  });

  afterAll(() => {
    vi.resetAllMocks();
  });

  const { TREE_DATA } = await vi.importActual<{ TREE_DATA: TreeNode[] }>("$core/__mocks__");

  it("throws an error when node is not found", () => {
    const emptyTree: TreeNode[] = [];

    expect(() => move(emptyTree, "1", null)).toThrow(new Error("[Categorizr:move] Cannot found the source node with the given id"));
    expect(() => move(TREE_DATA, "10", null)).toThrow(new Error("[Categorizr:move] Cannot found the source node with the given id"));
  });

  it("throws an error when try move node to its own descendants", () => {
    expect(() => move(TREE_DATA, "1", "3")).toThrow(new Error("[Categorizr:move] Cannot move the node into its own descendants."));
    expect(() => move(TREE_DATA, "1", "5")).toThrow(new Error("[Categorizr:move] Cannot move the node into its own descendants."));
    expect(() => move(TREE_DATA, "3", "5")).toThrow(new Error("[Categorizr:move] Cannot move the node into its own descendants."));
  });

  it("moved nodeId:4 to nodeId:3 should", () => {
    expect(move(TREE_DATA, "4", "3")).toStrictEqual([
      {
        id: "1",
        name: "category-1",
        children: [
          {
            id: "3",
            name: "sub-category-1",
            children: [
              {
                id: "5",
                name: "sub-category-3",
                children: [],
              },
              {
                id: "4",
                name: "sub-category-2",
                children: [],
              },
            ],
          },
        ],
      },
      {
        id: "2",
        name: "category-2",
        children: [],
      },
    ]);

    expect(move(TREE_DATA, "4", "3")).toMatchSnapshot();

    move(TREE_DATA, "4", "3", (newTree) => {
      expect(newTree).toMatchSnapshot();
    });
  });

  it("moved nodeId:3 to nodeId:null (first level depth) should", () => {
    expect(move(TREE_DATA, "3", null)).toStrictEqual([
      {
        id: "1",
        name: "category-1",
        children: [],
      },
      {
        id: "2",
        name: "category-2",
        children: [
          {
            id: "4",
            name: "sub-category-2",
            children: [],
          },
        ],
      },
      {
        id: "3",
        name: "sub-category-1",
        children: [
          {
            id: "5",
            name: "sub-category-3",
            children: [],
          },
        ],
      },
    ]);

    expect(move(TREE_DATA, "3", null)).toMatchSnapshot();

    move(TREE_DATA, "3", null, (newTree) => {
      expect(newTree).toMatchSnapshot();
    });
  });

  it("skip move processes if node moved to same dest", () => {
    expect(move(TREE_DATA, "3", "1")).toStrictEqual(TREE_DATA);
    expect(move(TREE_DATA, "2", null)).toStrictEqual(TREE_DATA);
  });
});
