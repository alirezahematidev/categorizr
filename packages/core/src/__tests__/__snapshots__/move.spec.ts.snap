// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`move > moved nodeId:3 to nodeId:null (first level depth) should 1`] = `
[
  {
    "children": [],
    "id": "1",
    "name": "category-1",
  },
  {
    "children": [
      {
        "children": [],
        "id": "4",
        "name": "sub-category-2",
      },
    ],
    "id": "2",
    "name": "category-2",
  },
  {
    "children": [
      {
        "children": [],
        "id": "5",
        "name": "sub-category-3",
      },
    ],
    "id": "3",
    "name": "sub-category-1",
  },
]
`;

exports[`move > moved nodeId:3 to nodeId:null (first level depth) should 2`] = `
[
  {
    "children": [],
    "id": "1",
    "name": "category-1",
  },
  {
    "children": [
      {
        "children": [],
        "id": "4",
        "name": "sub-category-2",
      },
    ],
    "id": "2",
    "name": "category-2",
  },
  {
    "children": [
      {
        "children": [],
        "id": "5",
        "name": "sub-category-3",
      },
    ],
    "id": "3",
    "name": "sub-category-1",
  },
]
`;

exports[`move > moved nodeId:4 to nodeId:3 should 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "id": "5",
            "name": "sub-category-3",
          },
          {
            "children": [],
            "id": "4",
            "name": "sub-category-2",
          },
        ],
        "id": "3",
        "name": "sub-category-1",
      },
    ],
    "id": "1",
    "name": "category-1",
  },
  {
    "children": [],
    "id": "2",
    "name": "category-2",
  },
]
`;

exports[`move > moved nodeId:4 to nodeId:3 should 2`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": [],
            "id": "5",
            "name": "sub-category-3",
          },
          {
            "children": [],
            "id": "4",
            "name": "sub-category-2",
          },
        ],
        "id": "3",
        "name": "sub-category-1",
      },
    ],
    "id": "1",
    "name": "category-1",
  },
  {
    "children": [],
    "id": "2",
    "name": "category-2",
  },
]
`;
